### PR TITLE
Remove upper password length bound.

### PR DESCRIPTION
**Pull request by Heinrich-XIAO**

### CODE REVIEW

**Comment by Heinrich-XIAO** [comment]
#20

**Comment by tsevasa** [comment]
You corrected the if-check but the error message on that line still says "'Password must be between 6-30 characters long'".

File: src/server/controllers/createaccountController.js [hunk 1]
```diff
@@ -251,7 +251,7 @@ const isValidEmail = function (string) {
 
 const doPasswordFormatChecks = function (password, res) {
     // First we check password length
-    if (password.length < 6 || password.length > 30) return res.status(400).json({ 'message': 'Password must be between 6-30 characters long'});
+    if (password.length < 6) return res.status(400).json({ 'message': 'Password must be between 6-30 characters long'});
```
Comment (Naviary2) [inline src/server/controllers/createaccountController.js:254]: (The message sent to the client also still says between 6-30)

**Comment by Naviary2** [comment]
Is an unlimited password size going to allow people to internal-DDOS the server by sending extremely long passwords? This regex is run for every password submitted:

```js
const isValidPassword = function (string) {
    const regex = /^[a-zA-Z0-9!@#$%^&*\?]+$/;
    if (regex.test(string) === true) return true;
    return false;
}
```
To me this doesn't look complicated, I think this would be fast still. But who has more intuition for regexes?

**Comment by Heinrich-XIAO** [comment]
Its O(n), so it can't be too bad.

File: src/client/scripts/createaccount.js [hunk 2]
```diff
@@ -167,7 +167,7 @@ passwordInputElement.addEventListener('input', (event) => { // When password fie
     if (formatError) {
         passwordError.textContent = 'Password is in an incorrect format';
     } else if (lengthError) {
-        passwordError.textContent = 'Password must be between 6-30 characters long';
+        passwordError.textContent = 'Password must be more than 6 characters long';
```
Comment (Naviary2) [inline src/client/scripts/createaccount.js:170]: Just a nitpick, but "atleast 6" 😆

**Comment by Heinrich-XIAO** [comment]
> Is an unlimited password size going to allow people to internal-DDOS the server by sending extremely long passwords? This regex is run for every password submitted:
> 
> ```js
> const isValidPassword = function (string) {
>     const regex = /^[a-zA-Z0-9!@#$%^&*\?]+$/;
>     if (regex.test(string) === true) return true;
>     return false;
> }
> ```
> 
> To me this doesn't look complicated, I think this would be fast still. But who has more intuition for regexes?
I can't get it to take more than 0.01 seconds on a github codespace.

**Comment by Naviary2** [comment]
Screen Shot 2024-07-11 at 2 54 06 PM

Also need to remove this attribute on the html element

**Comment by Heinrich-XIAO** [comment]
Why was that originally 40 if the max is 30?

**Comment by Naviary2** [comment]
That's a discrepancy!

Also I checked and the login.html has a max length of 40 as well, gotta remove that too

**Comment by Idonotus** [comment]
truely unlimited passwords may lead to DoS attacks so have a high limit like 1024 of 256

**Comment by Naviary2** [comment]
When I create an account with an extremely long password, when I login again, ANY password I enter that matches at least the first approximate 64 digits will log me in no matter if digits after those first 64 are incorrect.

Is this the bcrypt package that caps the password lengths? Is it even worth allowing unlimited password size if beyond 64 gives no more security?

**Comment by Heinrich-XIAO** [comment]
How long is the "extremely long password"? 100? 200? 1000? 10000? or just more than 64?

**Comment by Naviary2** [comment]
> How long is the "extremely long password"? 100? 200? 1000? 10000? or just more than 64?

Around 1000

**Comment by 5jiji** [comment]
> > How long is the "extremely long password"? 100? 200? 1000? 10000? or just more than 64?
> 
> Around 1000

That is extremely good, as my pasword manager limits me to make 128 chars long password (auto-generated) for some reason

**Comment by Naviary2** [comment]
If we're going to cap it, we should probably do so at exactly bcrypt's limit.

Screen Shot 2024-07-11 at 8 32 09 PM

**Comment by Naviary2** [comment]
> That is extremely good, as my pasword manager limits me to make 128 chars long password (auto-generated) for some reason

72 is not as good 😄 Which is the limit of bcrypt which we're using

**Comment by Naviary2** [comment]
I suggest cap at 72. But if they have less than 6 characers we only say "Password must be 6+ characters", and if they try to enter more than 72 we say "Password must be not more than 72 characters." Most people don't make passwords that long so we don't need to tell them the limit unless they exceed it

**Comment by 5jiji** [comment]
> Bcrypt hashed password and secrets have a 72 characters limit.

Is it me, or i'm reading that 72 is the max characters for a **hashed** password, and not a \*true\* password?

**Comment by Naviary2** [comment]
> Is it me, or i'm reading that 72 is the max characters for a **hashed** password, and not a *true* password?

All the hashes are exactly 60 characters you can check in members.json. I already tested out creating a long password and confirmed that any password over 72 characters essentially ignores what you put after that.

**Comment by this-is-not-available** [comment]
I think part of the hash is metadata (hash algorithm, cost) so the hash part of the hash is less than that.

**Comment by Heinrich-XIAO** [comment]
Does that mean that the limit should be 72 or that we need a different hashing library/algorithm?

**Comment by this-is-not-available** [comment]
It means that more passwords will share the same hash, if encoded similarly. A prime example of the pigeonhole principle.

**Comment by Naviary2** [comment]
Why should we search for an alternate package that allows infinite password length, when 99.9% of people don't make passwords that long, and, we'd still have to use bcrypt for testing passwords of users who created their account before the switch to the new package?

**Comment by Heinrich-XIAO** [comment]
So, the maximum should just be 72?

**Comment by Naviary2** [comment]
> So, the maximum should just be 72?

YesI think so!

**Comment by Naviary2** [comment]
Ready to be merged!

**Pull request merged**